crates/polars-arrow/src/array/binview/mod.rs

```diff
@@ -0,0 +1,276 @@
+    buffers
+        .iter()
+        .map(|buf| (buf.as_ptr(), buf.len()))
+        .collect()
+}
+
+impl<T: ViewType + ?Sized> BinaryViewArrayGeneric<T> {
+    /// # Safety
+    /// The caller must ensure
+    /// - the data is valid utf8 (if required)
+    /// - The offsets match the buffers.
+    pub unsafe fn new_unchecked(
+        data_type: ArrowDataType,
+        views: Buffer<u128>,
+        buffers: Vec<Buffer<u8>>,
+        validity: Option<Bitmap>,
+    ) -> Self {
+        let raw_buffers = buffers_into_raw(&buffers);
+        Self {
+            data_type,
+            views,
+            buffers,
+            raw_buffers,
+            validity,
+            phantom: Default::default(),
+        }
+    }
+
+    pub fn try_new(
+        data_type: ArrowDataType,
+        views: Buffer<u128>,
+        buffers: Vec<Buffer<u8>>,
+        validity: Option<Bitmap>,
+    ) -> PolarsResult<Self> {
+        if T::IS_UTF8 {
+            // check utf8?
+            todo!()
+        }
+        // traverse views and validate offsets?
+        if let Some(validity) = &validity {
+            polars_ensure!(validity.len()== views.len(), ComputeError: "validity mask length must match the number of values" )
+        }
+
+        let raw_buffers = buffers_into_raw(&buffers);
+        Ok(Self {
+            data_type,
+            views,
+            buffers,
+            raw_buffers,
+            validity,
+            phantom: Default::default(),
+        })
+    }
+
+    /// Creates an empty [`BinaryViewArrayGeneric`], i.e. whose `.len` is zero.
+    #[inline]
+    pub fn new_empty(data_type: ArrowDataType) -> Self {
+        unsafe { Self::new_unchecked(data_type, Buffer::new(), vec![], None) }
+    }
+
+    /// Returns a new null [`BinaryViewArrayGeneric`] of `length`.
+    #[inline]
+    pub fn new_null(data_type: ArrowDataType, length: usize) -> Self {
+        let validity = Some(Bitmap::new_zeroed(length));
+        unsafe { Self::new_unchecked(data_type, Buffer::zeroed(length), vec![], validity) }
+    }
+
+    /// Returns the element at index `i`
+    /// # Panics
+    /// iff `i >= self.len()`
+    #[inline]
+    pub fn value(&self, i: usize) -> &T {
+        assert!(i < self.len());
+        unsafe { self.value_unchecked(i) }
+    }
+
+    /// Returns the element at index `i`
+    /// # Safety
+    /// Assumes that the `i < self.len`.
+    #[inline]
+    pub unsafe fn value_unchecked(&self, i: usize) -> &T {
+        let v = *self.views.get_unchecked(i);
+        let len = v as u32;
+
+        // view layout:
+        // length: 4 bytes
+        // prefix: 4 bytes
+        // buffer_index: 4 bytes
+        // offset: 4 bytes
+
+        // inlined layout:
+        // length: 4 bytes
+        // data: 12 bytes
+
+        let bytes = if len <= 12 {
+            let ptr = self.views.as_ptr() as *const u8;
+            std::slice::from_raw_parts(ptr.add(i * 16 + 4), len as usize)
+        } else {
+            let prefix = (v >> 64) as u32;
```

Comment:
Prefix is the same as buffer idx?

Reply:
Nope, that isn't indeed. :/ Should be `>> 32`.

Reply:
And, we also don't need the prefix here. :sweat_smile: